mace/ops/detection_output.cc

```diff
@@ -0,0 +1,262 @@
+    for (int j = 0; j < num_prior; ++j) {
+      float confidence = conf_ptr[j * num_classes + i];
+      if (confidence > confidence_threshold) {
+        BBox c = {bboxes[0 + j * 4],
+                  bboxes[1 + j * 4],
+                  bboxes[2 + j * 4],
+                  bboxes[3 + j * 4],
+                  i,
+                  confidence};
+        class_bbox_rects.push_back(c);
+      }
+    }
+    std::sort(class_bbox_rects.begin(), class_bbox_rects.end(), cmp);
+
+    // apply nms
+    std::vector<BBox> sorted_boxes;
+    NmsSortedBboxes(class_bbox_rects, nms_threshold,
+                    std::min(top_k, static_cast<int>(class_bbox_rects.size())),
+                    &sorted_boxes);
+    // gather
+    bbox_rects->insert(bbox_rects->end(), sorted_boxes.begin(),
+                       sorted_boxes.end());
+  }
+
+  std::sort(bbox_rects->begin(), bbox_rects->end(), cmp);
+
+  // output
+  int num_detected = keep_top_k < static_cast<int>(bbox_rects->size())
+                         ? keep_top_k
+                         : static_cast<int>(bbox_rects->size());
+  bbox_rects->resize(num_detected);
+
+  return num_detected;
+}
+}  // namespace mace
+
+
+
+namespace mace {
+namespace ops {
+
+class DetectionOutput : public Operation {
+ public:
+  explicit DetectionOutput(OpConstructContext *context) : Operation(context) {}
+};
+
+template <RuntimeType D, class T>
+class DetectionOutputOp;
+
+template<class T>
+class DetectionOutputOp<RuntimeType::RT_CPU, T> : public DetectionOutput {
+ public:
+  explicit DetectionOutputOp(OpConstructContext *context)
+      : DetectionOutput(context),
+        num_classes_(Operation::GetOptionalArg<int>("num_classes", 1)),
+        nms_threshold_(Operation::GetOptionalArg<float>("nms_threshold", 0.3f)),
+        nms_top_k_(Operation::GetOptionalArg<int>("nms_top_k", 100)),
+        keep_top_k_(Operation::GetOptionalArg<int>("keep_top_k", 100)),
+        confidence_threshold_(
+            Operation::GetOptionalArg<float>("confidence_threshold", 0.05f)) {}
+
+  MaceStatus Run(OpContext *context) override {
+    MACE_UNUSED(context);
+    Tensor *output = this->Output(0);
+
+    auto *loc_t = this->Input(0);
+    auto *conf_t = this->Input(1);
+    auto *pbox_t = this->Input(2);
+
+
+    auto num_prior = loc_t->shape()[1] / 4;
+
+    MACE_CHECK(num_prior == conf_t->shape()[1] / num_classes_,
+               "conf tensor shape miss match");
+    MACE_CHECK(num_prior == pbox_t->shape()[2] / 4,
+               "prior box tensor shape miss match");
+
+    const float *loc_ptr = loc_t->data<float>();
+    const float *conf_ptr = conf_t->data<float>();
+    const float *pbox_ptr = pbox_t->data<float>();
+
+    std::vector<BBox> bbox_rects;
+
+    DetectionOutput_CLA(loc_ptr, conf_ptr, pbox_ptr, num_prior, num_classes_,
+                        nms_threshold_, nms_top_k_, keep_top_k_,
+                        confidence_threshold_, &bbox_rects);
+
+    output->Clear();
```

Comment:
Can output > clear() be deleted?

Reply:
No, necessory for clear previous inference output. `bbox_rects.size()` is not fixed

Reply:
I see 'output > Reset (output)' on line 220, If the shape of output is correct, the output data will cover the original data and the redundant data will not be accessed?